layer/input_buffer.cc

```diff
@@ -0,0 +1,212 @@
+class UnixMemMappedInputBufferImpl : public InputBuffer::InputBufferImplBase {
+ public:
+  static absl::StatusOr<std::unique_ptr<InputBuffer::InputBufferImplBase>>
+  Create(const std::string& path) {
+    UnixMemMappedInputBufferImpl res;
+    res.file_descriptor_ = open(path.c_str(), O_RDONLY);
+    if (res.file_descriptor_ == -1) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to open file for read: ", path));
+    }
+
+    struct stat s = {};
+    if (fstat(res.file_descriptor_, &s) != 0) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to stat file: ", path));
+    }
+    size_t size = s.st_size;
+    if (size == 0) {
+      // We cannot mmap empty files, but we can return an empty buffer without
+      // mmaping the underlying file.
+      return std::make_unique<UnixMemMappedInputBufferImpl>(std::move(res));
+    }
+
+    // This memory is read-only, but we do not use a const pointer, because
+    // munmap takes the memory to unmap through a void * pointer. No non-const
+    // pointers ever leave this class.
+    void* data = mmap(0, size, PROT_READ, MAP_PRIVATE, res.file_descriptor_, 0);
+    if (data == MAP_FAILED) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to mmap file: ", path));
+    }
+
+    res.buffer_start_ = reinterpret_cast<uint8_t*>(data);
+    res.buffer_size_ = size;
+    SPL_LOG(INFO) << "Mmapped file " << path << " sz: " << size;
+    return std::make_unique<UnixMemMappedInputBufferImpl>(std::move(res));
+  }
+
+  absl::Span<const uint8_t> GetBuffer() const override {
+    assert(file_descriptor_ != -1);
+    assert(buffer_start_ != MAP_FAILED);
+    assert(buffer_start_ || buffer_size_ == 0);
+    return absl::MakeConstSpan(buffer_start_, buffer_size_);
+  }
+
+  UnixMemMappedInputBufferImpl(UnixMemMappedInputBufferImpl&& other)
+      : file_descriptor_(other.file_descriptor_),
+        buffer_start_(other.buffer_start_),
+        buffer_size_(other.buffer_size_) {
+    other.file_descriptor_ = -1;
+    other.buffer_start_ = nullptr;
+    other.buffer_size_ = 0;
+  }
+
+  UnixMemMappedInputBufferImpl& operator=(
+      UnixMemMappedInputBufferImpl&& other) {
+    reset();
+    std::swap(file_descriptor_, other.file_descriptor_);
+    std::swap(buffer_start_, other.buffer_start_);
+    std::swap(buffer_size_, other.buffer_size_);
+    return *this;
+  }
+
+  UnixMemMappedInputBufferImpl(const UnixMemMappedInputBufferImpl&) = delete;
+  UnixMemMappedInputBufferImpl& operator=(const UnixMemMappedInputBufferImpl&) =
+      delete;
+
+  ~UnixMemMappedInputBufferImpl() override { reset(); }
+
+ private:
+  UnixMemMappedInputBufferImpl() = default;
+
+  void reset() {
+    if (buffer_start_) munmap(buffer_start_, buffer_size_);
+    if (file_descriptor_ != -1) close(file_descriptor_);
+
+    file_descriptor_ = -1;
+    buffer_start_ = nullptr;
+    buffer_size_ = 0;
+  }
+
+  int file_descriptor_ = -1;
+  uint8_t* buffer_start_ = nullptr;
```

Comment:
Nit: please add a short comment for `uint8_t* buffer_start_ = nullptr;` as well.